### PR TITLE
GPU Memory Spaces

### PR DESCRIPTION
**Pull request by MrBurmark**
# Support running with multiple Memory Spaces

Add a runtime option to change the Memory Space used for most allocations. This allows us to compare performance using different memory spaces.
Allocations with specific purposes like those used by reductions and temporary memory for library calls are always allocated with a specific memory space.

- This PR is a feature
- It does the following:
  - Adds runtime changeable memory spaces at the request of myself

## How this works
Each time a variant is run the memory is setup and torn down. If a variant, like a gpu variant, needs memory in a specific space it will make a copy in that space later when that variant is run. This adds extra code to copy memory around in the omp target, cuda, and hip kernel implementation files.
The main insight is that because memory is setup and torn down for each variant we can allocate memory for the variant that is about to be run so we don't have to move it later. This is accomplished by adding a layer of memory functions in KernelBase that get the memory space for the current variant and call the next layer in DataUtils that take the memory space and call the right allocation method.

### CODE REVIEW

File: src/common/RAJAPerfSuite.cpp [hunk 1]
```diff
@@ -371,6 +421,7 @@ const std::string& getVariantName(VariantID vid)
   return VariantNames[vid];
 }
 
+
```
Comment (MrBurmark) [inline src/common/RAJAPerfSuite.cpp:424]:
```suggestion
```

**Comment by MrBurmark** [comment]
For hip it probably makes sense to support fine grained and coarse grained memory, do you think it makes sense to support this as a part of the memory spaces or as a separate axis?

File: src/common/RAJAPerfSuite.hpp [hunk 2]
```diff
@@ -233,6 +233,63 @@ enum FeatureID {
...
+  Pinned,
+  PinnedFine,
+  PinnedCoarse,
+  Managed,
+  ManagedAdviseFine,
+  ManagedAdviseCoarse,
+  Device,
+  DeviceFine,
+
+  NumSpaces // Keep this one last and NEVER comment out (!!)
```
Comment (CRobeck) [inline src/common/RAJAPerfSuite.hpp:288]: Is there a reason we have NumSpaces in the enum and just not a constexpr member variable?

File: src/common/HipDataUtils.hpp [hunk 3]
```diff
@@ -133,6 +167,39 @@ void allocHipPinnedData(T& pptr, int len)
               hipHostMallocMapped ) );
 }
 
+/*!
```
Comment (CRobeck) [inline src/common/HipDataUtils.hpp:170]:
```suggestion
/*!
constexpr int getL1CacheLineSize()
{
     return 64;
}
```

Comment (CRobeck) [inline src/common/HipDataUtils.hpp:170]:
```suggestion
/*!

constexpr int getDefaultAlignment()
{    
    return 4096;
}
```

**Comment by MrBurmark** [comment]
@rhornung67 @artv3 I updated this PR, take a look. There are still some minor improvements that I would make like moving some of the routines into source files from headers but its mostly reasonable.

**Comment by rhornung67** [comment]
I need to study this to wrap my head around it. Probably will run in a debugger to see how this works.

File: src/common/KernelBase.hpp [hunk 4]
```diff
@@ -206,81 +209,166 @@ class KernelBase
...
   template <typename T>
-  void allocAndInitDataConst(T*& ptr, int len, T val,
-                             VariantID vid)
+  rajaperf::AutoDataMover<T> allocSetupData(T*& ptr, int len, VariantID vid)
   {
-    rajaperf::detail::allocAndInitDataConst(ptr, len, getDataAlignment(), val, vid);
+    rajaperf::allocData(getHostAccessibleDataSpace(vid),
+        ptr, len, getDataAlignment());
+    return {getDataSpace(vid), getHostAccessibleDataSpace(vid), ptr, len, getDataAlignment()};
   }
```
Comment (MrBurmark) [inline src/common/KernelBase.hpp:313]: But if there is any data initialization in the Kernel's setup method then you use the SetupData alloc methods which allocate memory that is accessible on the host, then you do setup, then the AutoDataMover moves it back at the end of the setup method. Maybe AutoDataMover is overly complicated, but I wanted the data to always be moved to the right DataSpace.

File: src/apps/DEL_DOT_VEC_2D.cpp [hunk 5]
```diff
@@ -73,10 +73,19 @@ void DEL_DOT_VEC_2D::setUp(VariantID vid, size_t RAJAPERF_UNUSED_ARG(tune_idx))
...
+  {
+    auto reset_x = scopedMoveData(m_x, m_array_length, vid);
+    auto reset_y = scopedMoveData(m_y, m_array_length, vid);
+    auto reset_rz = scopedMoveData(m_real_zones, m_domain->n_real_zones, vid);
+
+    Real_type dx = 0.2;
+    Real_type dy = 0.1;
+    setMeshPositions_2d(m_x, dx, m_y, dy, *m_domain);
+    setRealZones_2d(m_real_zones, *m_domain);
+  }
```
Comment (MrBurmark) [inline src/apps/DEL_DOT_VEC_2D.cpp:88]: But if there is any data initialization in the Kernel's setup method then you use the scopedMoveData method which reallocates the memory to be accessible on the host, then you do setup, then the AutoDataMover moves it back at the end of the scope. Maybe AutoDataMover is overly complicated, but I wanted the data to always be moved to the right DataSpace.

**Comment by MrBurmark** [comment]
I think I have gotten back to correctness across all the backends at this point.

**Review by rhornung67 — APPROVED** [review]
Looks good. Dev guide changes explaining the mechanics in the "adding a kernel" section in a separate PR?

**Comment by rhornung67** [comment]
@MrBurmark I've been restarting GitLab CI and merging PRs as checks come back green. I think this PR should be merged next. There is a conflict due to the merge of https://github.com/LLNL/RAJAPerf/pull/320. Do you want to resolve this or should I? Is there any more work you want to do on this PR before merging?

**Comment by rhornung67** [comment]
@MrBurmark looks good!

**Comment by MrBurmark** [comment]
I'll merge then

**Pull request merged**